File: split/writeout.pas
procedure writeout(p: halfword);
var
  oldsetting: 0..21;
  oldmode: integer;
  j: smallnumber;
  q, r: halfword;
begin
  {1371:}
q := getavail;
  mem[q].hh.lh := 637;
  r := getavail;
  mem[q].hh.rh := r;
  mem[r].hh.lh := 6717;
  begintokenlist(q, 4);
  begintokenlist(mem[p + 1].hh.rh, 15);
  q := getavail;
  mem[q].hh.lh := 379;
  begintokenlist(q, 4);
  oldmode := curlist.modefield;
  curlist.modefield := 0;
  curcs := writeloc;
  q := scantoks(false, true);
  gettoken;
  if curtok <> 6717 then{1372:}

  begin
    begin
      if interaction = 3 then;
      printnl(262);
      print(1296);
    end;
    begin
      helpptr := 2;
      helpline[1] := 1297;
      helpline[0] := 1011;
    end;
    error;
    repeat
      gettoken;
    until curtok = 6717;
  end{:1372}
;
  curlist.modefield := oldmode;
  endtokenlist{:1371}
;
  oldsetting := selector;
  j := mem[p + 1].hh.lh;
  if writeopen[j] then
    selector := j
  else
  begin
    if (j = 17) and (selector = 19) then
      selector := 18;
    printnl(338);
  end;
  tokenshow(defref);
  println;
  flushlist(defref);
  selector := oldsetting;
end;

File: cpp/src/writeout.cpp
#include "writeout.h"
#include "noeud.h"
#include "lecture.h"
#include "impression.h"
#include "erreur.h"

constexpr int end_write_token = cs_token_flag+end_write;

void writeout(NotOpenWriteWhatsitNode *p)
{
	auto q = new TokenNode(right_brace_token+'}');
	q->link = new TokenNode(end_write_token);
	ins_list(q);
	begintokenlist(p->write_tokens, write_text);
	ins_list(new TokenNode(left_brace_token+'{'));
	int oldmode = mode;
	mode = 0;
	Token t;
	t.cs = writeloc;
	q = scantoks(false, true, t);
	t = gettoken();
	if (t.tok != end_write_token)
	{
		error("Unbalanced write command", "On this page there's a \\write with fewer real {'s than }'s.\nI can't handle that very well; good luck.");
		do
			t = gettoken();
		while (t.tok != end_write_token);
	}
	mode = oldmode;
	endtokenlist();
	auto oldsetting = selector;
	auto j = p->write_stream;
	if (writeopen[j])
		selector = j;
	else
	{
		if (j == term_only && selector == term_and_log)
			selector = log_only;
		printnl("");
	}
	print(tokenshow(defref)+"\n");
	flushnodelist(defref);
	selector = oldsetting;
}
